File: Vjezbe05/spam-agent-frontend/src/app/models/api.models.ts
// ════════════════════════════════════════════════════════════════════════════════
//                     API RESPONSE MODELS
// ════════════════════════════════════════════════════════════════════════════════

export interface MessageDto {
  id: number;
  text: string;
  source: string;
  status: MessageStatus;
  trueLabel?: string;
  createdAtUtc: string;
  lastPrediction?: PredictionDto;
}

export interface PredictionDto {
  pSpam: number;
  decision: string;
  modelVersion: number;
  createdAtUtc: string;
}

export interface ModelVersionDto {
  id: number;
  version: number;
  trainTemplate: string;
  trainSetSize: number;
  goldIncludedCount: number;
  validationSetSize: number;
  metrics: MetricsDto;
  thresholdAllow: number;
  thresholdBlock: number;
  isActive: boolean;
  createdAtUtc: string;
}

export interface MetricsDto {
  accuracy: number;
  precision: number;
  recall: number;
  f1: number;
}

export interface SystemStatusDto {
  activeModel?: ModelVersionDto;
  settings: SettingsDto;
  queueStats: QueueStatsDto;
  datasetStats: DatasetStatsDto;
}

export interface SettingsDto {
  thresholdAllow: number;
  thresholdBlock: number;
  retrainGoldThreshold: number;
  newGoldSinceLastTrain: number;
  autoRetrainEnabled: boolean;
  lastRetrainAtUtc?: string;
}

export interface QueueStatsDto {
  queued: number;
  inInbox: number;
  inSpam: number;
  pendingReview: number;
  totalProcessed: number;
}

export interface DatasetStatsDto {
  totalMessages: number;
  uciMessages: number;
  runtimeMessages: number;
  trainPoolCount: number;
  validationCount: number;
  hamCount: number;
  spamCount: number;
  totalGoldLabels: number;
}

// ════════════════════════════════════════════════════════════════════════════════
//                     REQUEST MODELS
// ════════════════════════════════════════════════════════════════════════════════

export interface SendMessageRequest {
  text: string;
  source?: string;
}

export interface ReviewRequest {
  label: 'ham' | 'spam';
  note?: string;
  reviewedBy?: string;
}

export interface TrainRequest {
  template: 'Light' | 'Medium' | 'Full';
  activate: boolean;
}

export interface SettingsRequest {
  thresholdAllow?: number;
  thresholdBlock?: number;
  retrainGoldThreshold?: number;
  autoRetrainEnabled?: boolean;
  simulatorEnabled?: boolean;
  simulatorIntervalMs?: number;
  simulatorBatchSize?: number;
}

// ════════════════════════════════════════════════════════════════════════════════
//                     SIGNALR EVENT MODELS
// ════════════════════════════════════════════════════════════════════════════════

export interface MessageQueuedEvent {
  messageId: number;
  text: string;
  timestamp: string;
}

export interface MessageScoredEvent {
  messageId: number;
  textPreview: string;
  pSpam: number;
  decision: string;
  newStatus: string;
  trueLabel?: string;
  isCorrect?: boolean;
  timestamp: string;
}

export interface MessageMovedEvent {
  messageId: number;
  oldStatus: string;
  newStatus: string;
  label?: string;
  timestamp: string;
}

export interface ModelRetrainedEvent {
  newVersion: number;
  template: string;
  metrics: MetricsDto;
  isActivated: boolean;
  timestamp: string;
}

export interface StatsUpdatedEvent {
  queueStats: QueueStatsDto;
  newGoldSinceLastTrain: number;
  retrainGoldThreshold: number;
  timestamp: string;
}

// ════════════════════════════════════════════════════════════════════════════════
//                     ENUMS & TYPES
// ════════════════════════════════════════════════════════════════════════════════

export type MessageStatus = 'Queued' | 'InInbox' | 'InSpam' | 'PendingReview' | 'Archived';
export type SpamDecision = 'Allow' | 'PendingReview' | 'Block';
export type TrainTemplate = 'Light' | 'Medium' | 'Full';

// ════════════════════════════════════════════════════════════════════════════════
//                     UI MODELS
// ════════════════════════════════════════════════════════════════════════════════

export interface MessageCard extends MessageDto {
  isNew?: boolean;
  animationClass?: string;
}

export interface GoldProgress {
  current: number;
  threshold: number;
  percentage: number;
  willRetrain: boolean;
}

export interface SimulatorStatus {
  available: boolean;
  enabled: boolean;
  intervalMs: number;
  batchSize: number;
}

File: Vjezbe05/spam-agent-backend/AiAgents.SpamAgent.Web/Models/ApiModels.cs
/*
 * ═══════════════════════════════════════════════════════════════════════════════
 *          SPAM AGENT WEB - API MODELI (DTOs)
 * ═══════════════════════════════════════════════════════════════════════════════
 */

using System;
using System.Collections.Generic;
using AiAgents.SpamAgent.Domain;

namespace AiAgents.SpamAgent.Web.Models;

// ════════════════════════════════════════════════════════════════════════════════
//                     REQUEST MODELI
// ════════════════════════════════════════════════════════════════════════════════

/// <summary>
/// Request za slanje nove poruke u queue.
/// </summary>
public class SendMessageRequest
{
    public string Text { get; set; } = string.Empty;
    public string? Source { get; set; }
}

/// <summary>
/// Request za moderatorski review.
/// </summary>
public class ReviewRequest
{
    public string Label { get; set; } = string.Empty; // "ham" ili "spam"
    public string? Note { get; set; }
    public string? ReviewedBy { get; set; }
}

/// <summary>
/// Request za trening modela.
/// </summary>
public class TrainRequest
{
    public string Template { get; set; } = "Medium"; // Light, Medium, Full
    public bool Activate { get; set; } = true;
}

/// <summary>
/// Request za postavljanje pragova.
/// </summary>
public class ThresholdsRequest
{
    public double ThresholdAllow { get; set; }
    public double ThresholdBlock { get; set; }
}

/// <summary>
/// Request za postavke.
/// </summary>
public class SettingsRequest
{
    public double? ThresholdAllow { get; set; }
    public double? ThresholdBlock { get; set; }
    public int? RetrainGoldThreshold { get; set; }
    public bool? AutoRetrainEnabled { get; set; }
    public bool? SimulatorEnabled { get; set; }
    public int? SimulatorIntervalMs { get; set; }
    public int? SimulatorBatchSize { get; set; }
}

// ════════════════════════════════════════════════════════════════════════════════
//                     RESPONSE MODELI
// ════════════════════════════════════════════════════════════════════════════════

/// <summary>
/// Osnovna informacija o poruci.
/// </summary>
public class MessageDto
{
    public long Id { get; set; }
    public string Text { get; set; } = string.Empty;
    public string Source { get; set; } = string.Empty;
    public string Status { get; set; } = string.Empty;
    public string? TrueLabel { get; set; }
    public DateTime CreatedAtUtc { get; set; }
    public PredictionDto? LastPrediction { get; set; }
}

/// <summary>
/// Predikcija za poruku.
/// </summary>
public class PredictionDto
{
    public double PSpam { get; set; }
    public string Decision { get; set; } = string.Empty;
    public int ModelVersion { get; set; }
    public DateTime CreatedAtUtc { get; set; }
}

/// <summary>
/// Informacija o verziji modela.
/// </summary>
public class ModelVersionDto
{
    public int Id { get; set; }
    public int Version { get; set; }
    public string TrainTemplate { get; set; } = string.Empty;
    public int TrainSetSize { get; set; }
    public int GoldIncludedCount { get; set; }
    public int ValidationSetSize { get; set; }
    public MetricsDto Metrics { get; set; } = new();
    public double ThresholdAllow { get; set; }
    public double ThresholdBlock { get; set; }
    public bool IsActive { get; set; }
    public DateTime CreatedAtUtc { get; set; }
}

/// <summary>
/// Metrike modela.
/// </summary>
public class MetricsDto
{
    public double Accuracy { get; set; }
    public double Precision { get; set; }
    public double Recall { get; set; }
    public double F1 { get; set; }
}

/// <summary>
/// Status sistema.
/// </summary>
public class SystemStatusDto
{
    public ModelVersionDto? ActiveModel { get; set; }
    public SettingsDto Settings { get; set; } = new();
    public QueueStatsDto QueueStats { get; set; } = new();
    public DatasetStatsDto DatasetStats { get; set; } = new();
}

/// <summary>
/// Postavke sistema.
/// </summary>
public class SettingsDto
{
    public double ThresholdAllow { get; set; }
    public double ThresholdBlock { get; set; }
    public int RetrainGoldThreshold { get; set; }
    public int NewGoldSinceLastTrain { get; set; }
    public bool AutoRetrainEnabled { get; set; }
    public DateTime? LastRetrainAtUtc { get; set; }
}

/// <summary>
/// Statistika queue-a.
/// </summary>
public class QueueStatsDto
{
    public int Queued { get; set; }
    public int InInbox { get; set; }
    public int InSpam { get; set; }
    public int PendingReview { get; set; }
    public int TotalProcessed { get; set; }
}

/// <summary>
/// Statistika dataseta.
/// </summary>
public class DatasetStatsDto
{
    public int TotalMessages { get; set; }
    public int UciMessages { get; set; }
    public int RuntimeMessages { get; set; }
    public int TrainPoolCount { get; set; }
    public int ValidationCount { get; set; }
    public int HamCount { get; set; }
    public int SpamCount { get; set; }
    public int TotalGoldLabels { get; set; }
}

// ════════════════════════════════════════════════════════════════════════════════
//                     SIGNALR EVENT MODELI
// ════════════════════════════════════════════════════════════════════════════════

/// <summary>
/// Event kad se poruka doda u queue.
/// </summary>
public class MessageQueuedEvent
{
    public long MessageId { get; set; }
    public string Text { get; set; } = string.Empty;
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// Event kad se poruka scoruje.
/// </summary>
public class MessageScoredEvent
{
    public long MessageId { get; set; }
    public string TextPreview { get; set; } = string.Empty;
    public double PSpam { get; set; }
    public string Decision { get; set; } = string.Empty;
    public string NewStatus { get; set; } = string.Empty;
    public string? TrueLabel { get; set; }
    public bool? IsCorrect { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// Event kad se poruka premjesti (review).
/// </summary>
public class MessageMovedEvent
{
    public long MessageId { get; set; }
    public string OldStatus { get; set; } = string.Empty;
    public string NewStatus { get; set; } = string.Empty;
    public string? Label { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// Event kad se model retrenira.
/// </summary>
public class ModelRetrainedEvent
{
    public int NewVersion { get; set; }
    public string Template { get; set; } = string.Empty;
    public MetricsDto Metrics { get; set; } = new();
    public bool IsActivated { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// Event sa statistikom (periodično).
/// </summary>
public class StatsUpdatedEvent
{
    public QueueStatsDto QueueStats { get; set; } = new();
    public int NewGoldSinceLastTrain { get; set; }
    public int RetrainGoldThreshold { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}
